### PR TITLE
NODE-1868: Blockchain refactoring

### PR DESCRIPTION
**Pull request by phearnot**

### CODE REVIEW

**Comment by lazaridiscom** [comment]
Any more (public) info about this seemingly huge change available?

File: node/src/main/scala/com/wavesplatform/state/diffs/BalanceDiffValidation.scala [hunk 1]
```diff
@@ -62,8 +59,6 @@ object BalanceDiffValidation extends ScorexLogging {
     }
   }
 
-  private def negativeAssetsInfo(p: Portfolio): Map[ByteStr, Long] =
-    p.assets.collect {
-      case (asset, balance) if balance < 0 => (asset.id, balance)
-    }
+  private def negativeAssetsInfo(b: Blockchain, acc: Address, diff: Portfolio): Map[ByteStr, Long] =
```
Comment (Karasiq) [inline node/src/main/scala/com/wavesplatform/state/diffs/BalanceDiffValidation.scala]: Should it return only negative values?

File: node/src/test/scala/com/wavesplatform/api/common/AddressTransactionsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package com.wavesplatform.api.common
+
+import com.wavesplatform.WithDB
+import org.scalatest.FreeSpec
+
+class AddressTransactionsSpec extends FreeSpec with WithDB {
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/api/common/AddressTransactionsSpec.scala]: Not implemented

File: node/src/test/scala/com/wavesplatform/api/common/CommonAccountApiSpec.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package com.wavesplatform.api.common
+
+import org.scalatest.{FreeSpec, Matchers}
+
+class CommonAccountApiSpec extends FreeSpec with Matchers {
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/api/common/CommonAccountApiSpec.scala]: Not impemented

File: node/src/test/scala/com/wavesplatform/api/common/CommonTransactionsApiSpec.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.wavesplatform.api.common
+
+import org.scalatest.FreeSpec
+
+class CommonTransactionsApiSpec extends FreeSpec {
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/api/common/CommonTransactionsApiSpec.scala]: Not implemented

File: node/src/test/scala/com/wavesplatform/history/BlockRewardSpec.scala [hunk 5]
```diff
@@ -240,19 +240,22 @@ class BlockRewardSpec extends FreeSpec with ScalaCheckPropertyChecks with WithDo
 
           d.levelDBWriter.height shouldBe BlockRewardActivationHeight - 1
           d.levelDBWriter.balance(miner1.toAddress) shouldBe InitialMinerBalance + OneFee
-          d.levelDBWriter.totalFee(BlockRewardActivationHeight - 1) shouldBe OneTotalFee.some
+          // todo
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/history/BlockRewardSpec.scala]: Not implemented

File: node/src/test/scala/com/wavesplatform/history/BlockchainUpdaterKeyAndMicroBlockConflictTest.scala [hunk 6]
```diff
@@ -63,10 +63,12 @@ class BlockchainUpdaterKeyAndMicroBlockConflictTest
       case (genesisBlock, Seq(block1, block2), microBlocks, address) =>
         withDomain(DataAndMicroblocksActivatedAt0WavesSettings) { d =>
           Seq(genesisBlock, block1, block2).foreach(d.blockchainUpdater.processBlock(_) shouldBe 'right)
-          d.blockchainUpdater.accountDataKeys(address) shouldBe Set("test")
+          // todo
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/history/BlockchainUpdaterKeyAndMicroBlockConflictTest.scala]: Not implemented

File: node/src/test/scala/com/wavesplatform/http/ProtoVersionTransactionsSpec.scala [hunk 7]
```diff
@@ -43,22 +50,23 @@ class ProtoVersionTransactionsSpec extends RouteSpec("/transactions") with RestA
...
-  }
+  private val utxPoolSynchronizer: UtxPoolSynchronizer = mock[UtxPoolSynchronizer]
+
+  private val transactionsApi = mock[CommonTransactionsApi]
+  private val route: Route =
+    TransactionsApiRoute(restAPISettings, transactionsApi, testWallet, blockchain, Coeval(utx.size), utxPoolSynchronizer, ntpTime).route
 
-  private def route: Route = TransactionsApiRoute(restAPISettings, wallet, blockchain, utx, utxPoolSynchronizer, new TestTime).route
+  private def test(f: => Any): Unit = {
+    (utxPoolSynchronizer.publish _).expects(*).anyNumberOfTimes().returning(TracedResult(Right(true)))
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/http/ProtoVersionTransactionsSpec.scala]: For what its done?

File: node/src/test/scala/com/wavesplatform/utx/UtxPoolSpecification.scala [hunk 8]
```diff
@@ -554,27 +552,27 @@ class UtxPoolSpecification
...
-            }
-          }
-
-          val changedAssetIds = pessimisticAssetIds -- baseAssetIds
-          withClue("changed") {
-            changedAssetIds.foreach { assetId =>
-              basePortfolio.balanceOf(assetId) should not be basePortfolio.balanceOf(assetId)
-            }
-          }
+        // todo
```
Comment (Karasiq) [inline node/src/test/scala/com/wavesplatform/utx/UtxPoolSpecification.scala]: Not implemented

**Pull request merged**